### PR TITLE
Implement SushiTee's shell sort, rewrite it to match pseudocode; deprecate heap, quick, merge sorts

### PR DESCRIPTION
**Pull request by Dune-jr**
Implements @SushiTee's algorithms from https://github.com/teeworlds/teeworlds/pull/1169
- Deprecate heap, quick, merge sorts.
- Carefully rewrite the shell sort to match 1:1 the pseudocode in https://en.wikipedia.org/wiki/Shellsort (with small increments from @SushiTee's code)

`sort_range` is currently used for demos, country flags. `sort` is used in the editor.

### CODE REVIEW

**Comment by heinrich5991** [comment]
Permanent link to the Wikipedia article: [Shellsort#Pseudocode](https://en.wikipedia.org/w/index.php?title=Shellsort&oldid=857438671#Pseudocode).

File: src/base/tl/sorted_array.h [hunk 1]
```diff
@@ -28,11 +28,63 @@ class sorted_array : public array<T, ALLOCATOR>
...
+
+	/*
+		Function: sort_range_merge()
+			sorts the whole array using merge sort (stable)
+		Warning: needs testing
+	*/
+	/*void sort_range_merge()
+	{
+		sort_merge(all());
+	}*/
```
Comment (heinrich5991) [inline src/base/tl/sorted_array.h]: Remove code that is commented out IMO.

File: src/base/tl/algorithm.h [hunk 2]
```diff
@@ -79,38 +80,271 @@ R find_binary(R range, T value)
...
+	concept_backwarditeration::check(range);
+
+	// Sort an array a[0...n-1].
+	/* gaps = [...] */
+	const int gaps[] = {
+		1, 3, 7, 21, 48, 112,
+		336, 861, 1968, 4592, 13776,
+		33936, 86961, 198768, 463792, 1391376,
+		3402672, 8382192, 21479367, 49095696, 114556624,
+		343669872, 52913488, 2085837936};
```
Comment (heinrich5991) [inline src/base/tl/algorithm.h]: Where do these gaps come from? I only see the first few in the Wikipedia article.

File: src/base/tl/algorithm.h [hunk 3]
```diff
@@ -79,38 +80,271 @@ R find_binary(R range, T value)
...
+	typename R::type *front = &range.front(); // a[0]
+	typename R::type *back = &range.back(); // a[n]
+	// Start with the largest gap and work down to a gap of 1
+	/* foreach (gap in gaps) */
```
[the remark this answers is not in the document]
Comment (Dune-jr) [reply src/base/tl/algorithm.h]: I intended to somehow distinguish pseudocode from actual comments. Any other way you would fancy?

**Comment by Dune-jr** [comment]
> 1. I think we should move to a standard sorting function (either qsort, or rather C++ sort): They're correct and fast and perhaps most importantly _tested_. :)

Indeed, I think that is not worth the risk. :+1: 

> 2. If we don't do that perhaps the shell sort algorithm could be rewritten using indices like the pseudocode has them (at least one subtle bug was introduced by using pointers instead of indices).

Would be great, yeah, no matter how much I puzzled to make it match the pseudocode a bug still swept through

> 3. Otherwise, I left a few comments on the issues I saw.

I leave that on hold until it is decided whether the #1169 and #1989 should be scrapped away or not.

**Comment by oy** [comment]
fixed by 397e88e1d7073bd417455926fd0b7193d4067f13

**Pull request closed without merging**